subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c

```diff
@@ -0,0 +1,417 @@
+	net_buf_pull(pkt->frags, header_len);
+
+	r = zoap_packet_parse(&response, pkt);
+	if (r < 0) {
+		SYS_LOG_ERR("Invalid data received (err:%d)", r);
+		goto cleanup;
+	}
+
+	token = zoap_header_get_token(&response, &tkl);
+	pending = zoap_pending_received(&response, pendings,
+					NUM_PENDINGS);
+	if (pending) {
+		/* If necessary cancel retransmissions */
+	}
+
+	SYS_LOG_DBG("checking for reply from [%s]",
+		    sprint_ip_addr(&from_addr));
+	reply = zoap_response_received(&response, &from_addr,
+				       replies, NUM_REPLIES);
+	if (!reply) {
+		SYS_LOG_ERR("No handler for response");
+	} else {
+		SYS_LOG_DBG("reply handled reply:%p", reply);
+		zoap_reply_clear(reply);
+	}
+
+cleanup:
+	if (pkt) {
+		net_pkt_unref(pkt);
+	}
+}
+
+static void retransmit_request(struct k_work *work)
+{
+	struct zoap_pending *pending;
+	int r;
+
+	pending = zoap_pending_next_to_expire(pendings, NUM_PENDINGS);
+	if (!pending) {
+		return;
+	}
+
+	r = net_context_sendto(pending->pkt, &pending->addr,
+			       NET_SOCKADDR_MAX_SIZE,
+			       NULL, K_NO_WAIT, NULL, NULL);
+	if (r < 0) {
+		return;
+	}
+
+	if (!zoap_pending_cycle(pending)) {
+		zoap_pending_clear(pending);
+		return;
+	}
+
+	k_delayed_work_submit(&retransmit_work, pending->timeout);
+}
+
+static int transfer_request(struct zoap_block_context *ctx,
+			    const u8_t *token, u8_t tkl,
+			    zoap_reply_t reply_cb)
+{
+	struct zoap_packet request;
+	struct net_pkt *pkt = NULL;
+	struct zoap_pending *pending;
+	struct net_buf *frag;
+	struct zoap_reply *reply = NULL;
+	int ret;
+
+	/* send request */
+	pkt = net_pkt_get_tx(firmware_net_ctx, K_FOREVER);
+	if (!pkt) {
+		SYS_LOG_ERR("Unable to get TX packet, not enough memory.");
+		return -ENOMEM;
+	}
+
+	frag = net_pkt_get_data(firmware_net_ctx, K_FOREVER);
+	if (!frag) {
+		SYS_LOG_ERR("Unable to get DATA buffer, not enough memory.");
+		ret = -ENOMEM;
+		goto cleanup;
+	}
+
+	net_pkt_frag_add(pkt, frag);
+
+	ret = zoap_packet_init(&request, pkt);
+	if (ret < 0) {
+		SYS_LOG_ERR("zoap packet init error (err:%d)", ret);
+		return ret;
+	}
+
+	/* FIXME: Could be that zoap_packet_init() sets some defaults */
+	zoap_header_set_version(&request, 1);
+	zoap_header_set_type(&request, ZOAP_TYPE_CON);
+	zoap_header_set_code(&request, ZOAP_METHOD_GET);
+	zoap_header_set_id(&request, zoap_next_id());
+
+	if (token && tkl > 0) {
+		zoap_header_set_token(&request, token, tkl);
+	} else if (tkl == 0) {
+		zoap_header_set_token(&request, zoap_next_token(), 8);
+	}
+
+	/* set the reply handler */
+	if (reply_cb) {
+		reply = zoap_reply_next_unused(replies, NUM_REPLIES);
+		if (!reply) {
+			SYS_LOG_ERR("No resources for waiting for replies.");
+			ret = -ENOMEM;
+			goto cleanup;
+		}
+
+		zoap_reply_init(reply, &request);
+		reply->reply = reply_cb;
+	}
+
+	/* hard code URI path here -- should be pulled from package_uri */
+	ret = zoap_add_option(&request, ZOAP_OPTION_URI_PATH,
+			"large-create", strlen("large-create"));
```

Comment:
Done.  NOTE: This file will have some immediate follow-up patches once this PR is merged and the hard-coded location for URI path is removed.